# Introduction

This is a jupyter notebook version of the [yyg-seir-simulator](https://github.com/youyanggu/yyg-seir-simulator)'s `readme` file. You can run code and see documents from just inside the notebook.

We present the underlying SEIR model simulator behind the YYG / [covid19-projections.com](https://covid19-projections.com) model, as well a summarized set of parameters that helped generate the projections.

If your system supports Python, you can generate your own simulations in under 5 minutes. No prior Python experience is needed to run this tool. [Get started here](#setup).

![SEIR Model Diagram](https://upload.wikimedia.org/wikipedia/commons/3/3d/SEIR.PNG)


To begin, we want to be clear that this is **not** the full model used by [covid19-projections.com](https://covid19-projections.com). This is the underlying SEIR model without the machine learning layer to learn the parameters. In fact, **this simulator does not use any published data**: it only simulates infections, hospitalizations, and deaths given a single set of parameters. As a result, **this tool is meant to generate simulations, not projections**.


Unlike traditional SEIR models, our simulator does not use differential equations - we use an abstracted state machine that tracks the probability transitions between the 4 states of SEIR: **S**usceptible-**E**xposed-**I**nfectious-**R**ecovered/deceased. Learn more about how our SEIR model works on [our website](https://covid19-projections.com/model-details/).

Because this simulator has very little [dependencies](#dependencies) and does not rely on published data, it is fast to run, works right out of the box, and is easily modifiable. We've purposefully designed our simulator to be as lean and simple as possible.

The simulations produced by this program will not necessarily match the full model, but it is often be a close approximation. The full model for *covid19-projections.com* generates thousands of parameter sets for each region and weighs the parameters based on how the resulting simulations match the observed data. With that said, this is the full, unabridged SEIR model used to generate the simulations - no modifications have been made for this release.

In addition to the SEIR simulator, we provide the "best" set of parameters that our machine learning layer has learned to best fit the real-world observed data. This is done by taking a weighted mean (or median) of the individual parameters used in our full model. Because parameters are not independent, using only a single set of parameters may skew the simulation results when compared to the full model projections.

While this simulator may not be best suited to make projections (since it is not tuned on any published data), we believe this simulator is particular helpful for mapping out relative scenarios. For example, how much can we reduce infections/deaths if [people started social distancing 7 days earlier](#decrease-inflection-date-by-7-days). Or what if [20% of individuals self-quarantine after symptom onset](#simulate-effect-of-quarantine). Or if there was [no reopening](#assume-no-reopening).

# Dependencies

You need [Python 3](https://www.python.org/downloads/) and [NumPy](https://numpy.org/install/). That's it.

*Note: This was built and tested on Python 3.8, so it's possible that an older Python 3.x version may not be compatible.*


# Setup

1. Make sure you have [Python 3](https://www.python.org/downloads/) and [NumPy](https://numpy.org/install/) installed.
2. Clone our repository: ```git clone https://github.com/youyanggu/yyg-seir-simulator.git```
3. You are now ready to run our simulator! See sample usage below.

# Usage

To view a list of supported countries, states, and subregions, check [covid19-projections.com](https://covid19-projections.com/#view-projections).


## List all command line flags/features:

In [1]:
!python run_simulation.py --help

usage: run_simulation.py [-h] [--skip_hospitalizations]
                         [--quarantine_perc QUARANTINE_PERC]
                         [--quarantine_effectiveness QUARANTINE_EFFECTIVENESS]
                         [--save_csv_fname SAVE_CSV_FNAME]
                         [--simulation_start_date SIMULATION_START_DATE]
                         [--simulation_end_date SIMULATION_END_DATE]
                         [--best_params_dir BEST_PARAMS_DIR]
                         [--best_params_type {mean,median,top,top10}]
                         [--set_param SET_PARAM SET_PARAM]
                         [--change_param CHANGE_PARAM CHANGE_PARAM]
                         [--country COUNTRY] [--region REGION]
                         [--subregion SUBREGION] [-v]

Script to run simulations using the YYG/C19Pro SEIR model. Example: `python
run_simulation.py -v --best_params_dir best_params/latest --country US
--region CA`

optional arguments:
  -h, --help            show this help message

# Basic Usage

Note: `-v` means verbose mode. Remove the flag and you will get fewer print statements.

## US simulation:

In [2]:
!python run_simulation.py -v --best_params_dir best_params/latest --country US

YYG/C19PRO Simulator
Current time: 2020-07-25 16:44:46.933476
Loading params file: best_params/latest\US_ALL.json
best params type: mean
US |  | 
Parameters:
INITIAL_R_0               : 2.3027814492609235
LOCKDOWN_R_0              : 0.9163095222616071
INFLECTION_DAY            : 2020-03-17
RATE_OF_INFLECTION        : 0.33297885944808253
LOCKDOWN_FATIGUE          : 1.0
DAILY_IMPORTS             : 661.6780489462355
MORTALITY_RATE            : 0.01
REOPEN_DATE               : 2020-05-10
REOPEN_SHIFT_DAYS         : 0.0
REOPEN_R                  : 1.2322034012756486
REOPEN_INFLECTION         : 0.29880982541997025
POST_REOPEN_EQUILIBRIUM_R : 1.0021645755311086
FALL_R_MULTIPLIER         : 1.0010021401832796
--------------------------
Running simulation...
--------------------------
2020-02-05 - New / total infections: 662 / 662 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 2.303 - IFR: 1.00%
2020-02-06 - New / total infections: 662 / 1,323 - Hospital beds in use: 0 - New / t

## US state simulations:

In [3]:
!python run_simulation.py -v --best_params_dir best_params/latest --country US --region NY

YYG/C19PRO Simulator
Current time: 2020-07-25 16:45:32.633613
Loading params file: best_params/latest\US_NY.json
best params type: mean
US | NY | 
Parameters:
INITIAL_R_0               : 3.887947058301361
LOCKDOWN_R_0              : 0.7737024537501702
INFLECTION_DAY            : 2020-03-12
RATE_OF_INFLECTION        : 0.3683438419307332
LOCKDOWN_FATIGUE          : 1.1
DAILY_IMPORTS             : 630.7115662807779
MORTALITY_RATE            : 0.0125
REOPEN_DATE               : 2020-06-01
REOPEN_SHIFT_DAYS         : 0.0
REOPEN_R                  : 1.158156996557596
REOPEN_INFLECTION         : 0.2994429163439425
POST_REOPEN_EQUILIBRIUM_R : 1.001613419195743
FALL_R_MULTIPLIER         : 1.0010531870310975
--------------------------
Running simulation...
--------------------------
2020-02-16 - New / total infections: 631 / 631 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 3.888 - IFR: 1.25%
2020-02-17 - New / total infections: 631 / 1,261 - Hospital beds in use: 0 - New / tot

## US county simulations:

In [4]:
!python run_simulation.py -v --best_params_dir best_params/latest --country US --region CA --subregion "Los Angeles"

YYG/C19PRO Simulator
Current time: 2020-07-25 16:46:04.548046
Loading params file: best_params/latest/subregion\US_CA_Los-Angeles.json
best params type: mean
US | CA | Los Angeles
Parameters:
INITIAL_R_0               : 2.4208580312931094
LOCKDOWN_R_0              : 1.0183676782608169
INFLECTION_DAY            : 2020-03-13
RATE_OF_INFLECTION        : 0.3937520939347715
LOCKDOWN_FATIGUE          : 0.9999999999999999
DAILY_IMPORTS             : 130.7780428119251
MORTALITY_RATE            : 0.007499999999999998
REOPEN_DATE               : 2020-06-05
REOPEN_SHIFT_DAYS         : -14.0
REOPEN_R                  : 1.1734347317235652
REOPEN_INFLECTION         : 0.29999999999999993
POST_REOPEN_EQUILIBRIUM_R : 1.0014623076359332
FALL_R_MULTIPLIER         : 1.0008474207935716
--------------------------
Running simulation...
--------------------------
2020-02-16 - New / total infections: 131 / 131 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 2.421 - IFR: 0.75%
2020-02-17 - New /

## Global country simulations:

In [5]:
!python run_simulation.py -v --best_params_dir best_params/latest --country Brazil

YYG/C19PRO Simulator
Current time: 2020-07-25 16:46:33.139005
Loading params file: best_params/latest/global\Brazil_ALL.json
best params type: mean
Brazil | ALL | 
Parameters:
INITIAL_R_0               : 1.5529910022167281
LOCKDOWN_R_0              : 1.0422776690279731
INFLECTION_DAY            : 2020-04-18
RATE_OF_INFLECTION        : 0.21043277617623535
LOCKDOWN_FATIGUE          : 0.9999999999999998
DAILY_IMPORTS             : 774.2961425338387
MORTALITY_RATE            : 0.007499999999999998
REOPEN_DATE               : 2020-06-15
REOPEN_SHIFT_DAYS         : 0.0
REOPEN_R                  : 1.0986950004401237
REOPEN_INFLECTION         : 0.09999999999999998
POST_REOPEN_EQUILIBRIUM_R : 1.0024103497779249
FALL_R_MULTIPLIER         : 0.9999999999999998
--------------------------
Running simulation...
--------------------------
2020-02-22 - New / total infections: 774 / 774 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 1.553 - IFR: 0.75%
2020-02-23 - New / total infections

2020-05-01 - New / total infections: 164,404 / 3,772,455 - Hospital beds in use: 36,773 - New / total deaths: 195 / 1,911 - Mean R: 1.056 - IFR: 0.75%
2020-05-02 - New / total infections: 165,678 / 3,938,133 - Hospital beds in use: 38,914 - New / total deaths: 214 / 2,125 - Mean R: 1.050 - IFR: 0.75%
2020-05-03 - New / total infections: 166,842 / 4,104,975 - Hospital beds in use: 41,020 - New / total deaths: 234 / 2,360 - Mean R: 1.044 - IFR: 0.75%
2020-05-04 - New / total infections: 167,910 / 4,272,885 - Hospital beds in use: 43,070 - New / total deaths: 256 / 2,616 - Mean R: 1.040 - IFR: 0.75%
2020-05-05 - New / total infections: 168,895 / 4,441,780 - Hospital beds in use: 45,045 - New / total deaths: 278 / 2,894 - Mean R: 1.036 - IFR: 0.75%
2020-05-06 - New / total infections: 169,804 / 4,611,584 - Hospital beds in use: 46,927 - New / total deaths: 301 / 3,195 - Mean R: 1.032 - IFR: 0.75%
2020-05-07 - New / total infections: 170,648 / 4,782,232 - Hospital beds in use: 48,702 - New 

# Advanced Usage

## Save outputs to CSV file

In [7]:
!python run_simulation.py -v --best_params_dir best_params/latest --country US  --save_csv_fname us_simulation.csv

YYG/C19PRO Simulator
Current time: 2020-07-25 16:47:35.836809
Loading params file: best_params/latest\US_ALL.json
best params type: mean
US |  | 
Parameters:
INITIAL_R_0               : 2.3027814492609235
LOCKDOWN_R_0              : 0.9163095222616071
INFLECTION_DAY            : 2020-03-17
RATE_OF_INFLECTION        : 0.33297885944808253
LOCKDOWN_FATIGUE          : 1.0
DAILY_IMPORTS             : 661.6780489462355
MORTALITY_RATE            : 0.01
REOPEN_DATE               : 2020-05-10
REOPEN_SHIFT_DAYS         : 0.0
REOPEN_R                  : 1.2322034012756486
REOPEN_INFLECTION         : 0.29880982541997025
POST_REOPEN_EQUILIBRIUM_R : 1.0021645755311086
FALL_R_MULTIPLIER         : 1.0010021401832796
--------------------------
Running simulation...
--------------------------
2020-02-05 - New / total infections: 662 / 662 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 2.303 - IFR: 1.00%
2020-02-06 - New / total infections: 662 / 1,323 - Hospital beds in use: 0 - New / t

## Use a custom end date

In [9]:
!python run_simulation.py -v --best_params_dir best_params/latest --country US --simulation_end_date 2020-11-01


YYG/C19PRO Simulator
Current time: 2020-07-25 16:47:59.092224
Loading params file: best_params/latest\US_ALL.json
best params type: mean
US |  | 
Parameters:
INITIAL_R_0               : 2.3027814492609235
LOCKDOWN_R_0              : 0.9163095222616071
INFLECTION_DAY            : 2020-03-17
RATE_OF_INFLECTION        : 0.33297885944808253
LOCKDOWN_FATIGUE          : 1.0
DAILY_IMPORTS             : 661.6780489462355
MORTALITY_RATE            : 0.01
REOPEN_DATE               : 2020-05-10
REOPEN_SHIFT_DAYS         : 0.0
REOPEN_R                  : 1.2322034012756486
REOPEN_INFLECTION         : 0.29880982541997025
POST_REOPEN_EQUILIBRIUM_R : 1.0021645755311086
FALL_R_MULTIPLIER         : 1.0010021401832796
--------------------------
Running simulation...
--------------------------
2020-02-05 - New / total infections: 662 / 662 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 2.303 - IFR: 1.00%
2020-02-06 - New / total infections: 662 / 1,323 - Hospital beds in use: 0 - New / t

2020-06-18 - New / total infections: 268,288 / 21,513,969 - Hospital beds in use: 53,738 - New / total deaths: 626 / 121,245 - Mean R: 1.191 - IFR: 0.46%
2020-06-19 - New / total infections: 277,715 / 21,791,683 - Hospital beds in use: 55,527 - New / total deaths: 617 / 121,862 - Mean R: 1.190 - IFR: 0.44%
2020-06-20 - New / total infections: 287,369 / 22,079,053 - Hospital beds in use: 57,424 - New / total deaths: 610 / 122,473 - Mean R: 1.188 - IFR: 0.43%
2020-06-21 - New / total infections: 297,227 / 22,376,280 - Hospital beds in use: 59,423 - New / total deaths: 605 / 123,078 - Mean R: 1.187 - IFR: 0.42%
2020-06-22 - New / total infections: 307,259 / 22,683,539 - Hospital beds in use: 61,521 - New / total deaths: 601 / 123,679 - Mean R: 1.185 - IFR: 0.41%
2020-06-23 - New / total infections: 317,417 / 23,000,955 - Hospital beds in use: 63,715 - New / total deaths: 598 / 124,277 - Mean R: 1.182 - IFR: 0.40%
2020-06-24 - New / total infections: 327,631 / 23,328,586 - Hospital beds in

## Use a custom start date

In [10]:
!python run_simulation.py -v --best_params_dir best_params/latest --country US --simulation_start_date 2020-02-01


YYG/C19PRO Simulator
Current time: 2020-07-25 16:48:25.988040
Loading params file: best_params/latest\US_ALL.json
best params type: mean
US |  | 
Parameters:
INITIAL_R_0               : 2.3027814492609235
LOCKDOWN_R_0              : 0.9163095222616071
INFLECTION_DAY            : 2020-03-17
RATE_OF_INFLECTION        : 0.33297885944808253
LOCKDOWN_FATIGUE          : 1.0
DAILY_IMPORTS             : 661.6780489462355
MORTALITY_RATE            : 0.01
REOPEN_DATE               : 2020-05-10
REOPEN_SHIFT_DAYS         : 0.0
REOPEN_R                  : 1.2322034012756486
REOPEN_INFLECTION         : 0.29880982541997025
POST_REOPEN_EQUILIBRIUM_R : 1.0021645755311086
FALL_R_MULTIPLIER         : 1.0010021401832796
--------------------------
Running simulation...
--------------------------
2020-02-01 - New / total infections: 662 / 662 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 2.303 - IFR: 1.00%
2020-02-02 - New / total infections: 662 / 1,323 - Hospital beds in use: 0 - New / t

## Simulate effect of quarantine

In this scenario, we show how the course of the epidemic would be different if just 20% of infected individuals immediately self-quarantine after showing symptoms, reducing their own transmission by 25%. For the remaining 80% of infected individuals, we assume normal transmission. You can see a graph of this scenario [here](https://covid19-projections.com/us-self-quarantine).

This can also be used as a proxy to simulate the effect of mask-wearing: if 20% of individuals wear masks, thereby reducing the overall transmission rate by 25%.

In [11]:
!python run_simulation.py -v --best_params_dir best_params/latest --country US --quarantine_perc 0.2 --quarantine_effectiveness 0.25


YYG/C19PRO Simulator
Current time: 2020-07-25 16:49:07.608883
Loading params file: best_params/latest\US_ALL.json
best params type: mean
Quarantine percentage: 20%
Quarantine effectiveness: 25%
US |  | 
Parameters:
INITIAL_R_0               : 2.3027814492609235
LOCKDOWN_R_0              : 0.9163095222616071
INFLECTION_DAY            : 2020-03-17
RATE_OF_INFLECTION        : 0.33297885944808253
LOCKDOWN_FATIGUE          : 1.0
DAILY_IMPORTS             : 661.6780489462355
MORTALITY_RATE            : 0.01
REOPEN_DATE               : 2020-05-10
REOPEN_SHIFT_DAYS         : 0.0
REOPEN_R                  : 1.2322034012756486
REOPEN_INFLECTION         : 0.29880982541997025
POST_REOPEN_EQUILIBRIUM_R : 1.0021645755311086
FALL_R_MULTIPLIER         : 1.0010021401832796
--------------------------
Running simulation...
--------------------------
2020-02-05 - New / total infections: 662 / 662 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 2.303 - IFR: 1.00%
2020-02-06 - New / total in

## Use the top 10 best parameters rather than the mean

The four choices are: mean (default), median, top, top10


In [12]:
!python run_simulation.py -v --best_params_dir best_params/latest --country US --best_params_type top10


YYG/C19PRO Simulator
Current time: 2020-07-25 16:49:56.545208
Loading params file: best_params/latest\US_ALL.json
best params type: top10
US |  | 
Parameters:
INITIAL_R_0               : 2.400000000000001
LOCKDOWN_R_0              : 0.9500000000000001
INFLECTION_DAY            : 2020-03-14
RATE_OF_INFLECTION        : 0.40000000000000013
LOCKDOWN_FATIGUE          : 1.0
DAILY_IMPORTS             : 750.0
MORTALITY_RATE            : 0.010000000000000002
REOPEN_DATE               : 2020-05-09
REOPEN_SHIFT_DAYS         : 0.0
REOPEN_R                  : 1.155
REOPEN_INFLECTION         : 0.305
POST_REOPEN_EQUILIBRIUM_R : 1.0026593250745395
FALL_R_MULTIPLIER         : 1.0010678081150184
--------------------------
Running simulation...
--------------------------
2020-02-05 - New / total infections: 750 / 750 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 2.400 - IFR: 1.00%
2020-02-06 - New / total infections: 750 / 1,500 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Me

## Setting Parameters

We also provide support to customize your own parameters using the `--set_param` flag. The flag takes a pair of values: the name of the parameter to be changed and its value. You can use the flag multiple times.

See the available parameter options in the [next section](#parameters).

### Set initial R_0 to 2.0

In [13]:
!python run_simulation.py -v --best_params_dir best_params/latest --country US --set_param INITIAL_R_0 2.0

YYG/C19PRO Simulator
Current time: 2020-07-25 16:51:08.525061
Loading params file: best_params/latest\US_ALL.json
best params type: mean
---------------------------------------
Overwriting params from command line...
Setting INITIAL_R_0 to: 2.0
US |  | 
Parameters:
INITIAL_R_0               : 2.0
LOCKDOWN_R_0              : 0.9163095222616071
INFLECTION_DAY            : 2020-03-17
RATE_OF_INFLECTION        : 0.33297885944808253
LOCKDOWN_FATIGUE          : 1.0
DAILY_IMPORTS             : 661.6780489462355
MORTALITY_RATE            : 0.01
REOPEN_DATE               : 2020-05-10
REOPEN_SHIFT_DAYS         : 0.0
REOPEN_R                  : 1.2322034012756486
REOPEN_INFLECTION         : 0.29880982541997025
POST_REOPEN_EQUILIBRIUM_R : 1.0021645755311086
FALL_R_MULTIPLIER         : 1.0010021401832796
--------------------------
Running simulation...
--------------------------
2020-02-05 - New / total infections: 662 / 662 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 2.000 - IF

2020-05-27 - New / total infections: 57,590 / 7,305,361 - Hospital beds in use: 21,876 - New / total deaths: 456 / 43,194 - Mean R: 1.108 - IFR: 0.66%
2020-05-28 - New / total infections: 59,422 / 7,364,783 - Hospital beds in use: 21,498 - New / total deaths: 446 / 43,640 - Mean R: 1.128 - IFR: 0.66%
2020-05-29 - New / total infections: 61,434 / 7,426,218 - Hospital beds in use: 21,139 - New / total deaths: 435 / 44,076 - Mean R: 1.146 - IFR: 0.66%
2020-05-30 - New / total infections: 63,614 / 7,489,832 - Hospital beds in use: 20,801 - New / total deaths: 425 / 44,501 - Mean R: 1.161 - IFR: 0.65%
2020-05-31 - New / total infections: 65,950 / 7,555,782 - Hospital beds in use: 20,489 - New / total deaths: 415 / 44,916 - Mean R: 1.174 - IFR: 0.65%
2020-06-01 - New / total infections: 68,435 / 7,624,217 - Hospital beds in use: 20,208 - New / total deaths: 406 / 45,322 - Mean R: 1.184 - IFR: 0.65%
2020-06-02 - New / total infections: 71,061 / 7,695,278 - Hospital beds in use: 19,962 - New /

## Changing parameters

Sometimes you don't want to replace an existing parameter, but rather just change it. The `--change_param` flag takes a pair of values: the name of the parameter to be changed and the amount to change the value by. You can use the flag multiple times.

See the available parameter options in the [next section](#parameters).

### Increase initial R_0 by 0.1

In [14]:
!python run_simulation.py -v --best_params_dir best_params/latest --country US --change_param INITIAL_R_0 0.1


YYG/C19PRO Simulator
Current time: 2020-07-25 16:52:05.347089
Loading params file: best_params/latest\US_ALL.json
best params type: mean
---------------------------------------
Changing params from command line...
Increasing INITIAL_R_0 from 2.3027814492609235 to 2.4027814492609236
US |  | 
Parameters:
INITIAL_R_0               : 2.4027814492609236
LOCKDOWN_R_0              : 0.9163095222616071
INFLECTION_DAY            : 2020-03-17
RATE_OF_INFLECTION        : 0.33297885944808253
LOCKDOWN_FATIGUE          : 1.0
DAILY_IMPORTS             : 661.6780489462355
MORTALITY_RATE            : 0.01
REOPEN_DATE               : 2020-05-10
REOPEN_SHIFT_DAYS         : 0.0
REOPEN_R                  : 1.2322034012756486
REOPEN_INFLECTION         : 0.29880982541997025
POST_REOPEN_EQUILIBRIUM_R : 1.0021645755311086
FALL_R_MULTIPLIER         : 1.0010021401832796
--------------------------
Running simulation...
--------------------------
2020-02-05 - New / total infections: 662 / 662 - Hospital beds in us

2020-07-26 - New / total infections: 370,582 / 43,354,631 - Hospital beds in use: 169,089 - New / total deaths: 1,191 / 191,073 - Mean R: 0.936 - IFR: 0.30%
2020-07-27 - New / total infections: 365,554 / 43,720,185 - Hospital beds in use: 168,177 - New / total deaths: 1,207 / 192,280 - Mean R: 0.935 - IFR: 0.30%
2020-07-28 - New / total infections: 360,547 / 44,080,732 - Hospital beds in use: 167,033 - New / total deaths: 1,220 / 193,499 - Mean R: 0.935 - IFR: 0.30%
2020-07-29 - New / total infections: 355,565 / 44,436,297 - Hospital beds in use: 165,702 - New / total deaths: 1,231 / 194,730 - Mean R: 0.934 - IFR: 0.30%
2020-07-30 - New / total infections: 350,607 / 44,786,904 - Hospital beds in use: 164,224 - New / total deaths: 1,239 / 195,969 - Mean R: 0.933 - IFR: 0.30%
2020-07-31 - New / total infections: 345,677 / 45,132,580 - Hospital beds in use: 162,631 - New / total deaths: 1,245 / 197,215 - Mean R: 0.933 - IFR: 0.30%
2020-08-01 - New / total infections: 340,775 / 45,473,355 

## Using Your Own Parameters

Don't want to use our parameter set? Want to run a simulation for a region/country we do not support? No problem! You can simply modify `run_simulations.py` (in the `main()` function) and specify your own default parameters (list of parameters [below](#parameters)). Afterwards, you can simply run:

In [15]:
!python run_simulation.py -v


YYG/C19PRO Simulator
Current time: 2020-07-25 16:53:23.822116
None | ALL | 
Parameters:
INITIAL_R_0               : 2.24
LOCKDOWN_R_0              : 0.9
INFLECTION_DAY            : 2020-03-18
RATE_OF_INFLECTION        : 0.25
LOCKDOWN_FATIGUE          : 1.0
DAILY_IMPORTS             : 500
MORTALITY_RATE            : 0.01
REOPEN_DATE               : 2020-05-20
REOPEN_SHIFT_DAYS         : 0
REOPEN_R                  : 1.2
REOPEN_INFLECTION         : 0.3
POST_REOPEN_EQUILIBRIUM_R : 1.0
FALL_R_MULTIPLIER         : 1.001
--------------------------
Running simulation...
--------------------------
2020-02-01 - New / total infections: 500 / 500 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 2.240 - IFR: 1.00%
2020-02-02 - New / total infections: 500 / 1,000 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 2.240 - IFR: 1.00%
2020-02-03 - New / total infections: 500 / 1,500 - Hospital beds in use: 0 - New / total deaths: 0 / 0 - Mean R: 2.240 - IFR: 1.00%
2020-02-0

Note: You can still use the `--set_param` and `--change_param` flags from above, as well as the other flags explained in [Advanced Usage](#advanced-usage).

You are also welcome to change the default parameters in [`fixed_params.py`](fixed_params.py), though we would recommend caution when doing so as many of those parameters have already been optimized.

# Parameters

There are several parameters that we need to provide the simulator before it can run. We describe them below:

#### `INITIAL_R_0`

This is the initial basic reproduction number (R_0). This value is usually between 0.8-6. Read more about our R_t estimates [here](https://covid19-projections.com/about/#effective-reproduction-number-r).

#### `LOCKDOWN_R_0`

This is the post-mitigation effective reproduction number (R_t). This value is usually between 0.3-1.5.

#### `INFLECTION_DAY`

This is the date of the inflection point when `INITIAL_R_0` transitions to `LOCKDOWN_R_0`. This is usually an indicator of when people in a region began social distancing. For example, in New York, the inflection day is around [March 14](https://twitter.com/youyanggu/status/1262881629376180227).

#### `RATE_OF_INFLECTION`

This is the rate at which the `INITIAL_R_0` transitions to `LOCKDOWN_R_0`. A number closer to 1 indicates a faster transition (i.e. faster lockdown), while a number closer to 0 indicates a slower transition. For most region, this value is between 0.15-0.5. The more localized a region, the higher the rate of inflection. So for example, we estimate New York City has a `RATE_OF_INFLECTION` between 0.4-0.5, while the US as an entire country has a `RATE_OF_INFLECTION` of approximately 0.2-0.3.

#### `LOCKDOWN_FATIGUE`

We incorporate a lockdown fatigue multiplier that is applied to the R_t. This is to simulate the increase in mobility after several weeks of lockdown, despite the order not being lifted. The default value is 1. If this value is greater than 1, then it will contribute to an increase in infections in the weeks following the lockdown/mitigation. We do not use this for the majority of our projections.

#### `DAILY_IMPORTS`

To begin our simulation, we must initialize / bootstrap our model with an initial number of infected individuals. Any epidemic in a region must begin with a number of imported individuals. This value gradually goes to 0 as community spread overtakes imported cases as the major source of spread. This value typically ranges from 10-1000.

#### `MORTALITY_RATE`

This is the initial estimate of the infection fatality rate (IFR). This value is usually between 0.005-0.0125. Read more about our IFR estimates [here](https://covid19-projections.com/about/#infection-fatality-rate-ifr).

#### `REOPEN_DATE`

This is the date we estimate the region to reopen. Read more about our reopening assumptions [here](https://covid19-projections.com/about/#social-distancing).

#### `REOPEN_SHIFT_DAYS`

Even though some regions open on the same date, infections can begin increasing faster or slower than we'd normally expect. We use the `REOPEN_SHIFT_DAYS` to account for that. For example, a value of 7 means that we are shifting the `REOPEN_DATE` to be 7 days later, while a value of -7 means we are shifting it to be 7 days earlier. The default value is 0.

#### `REOPEN_R`

(Added 2020-07-22) This is the maximum reopening R_t value. It takes roughly one month after the reopening to reach this value.

#### `REOPEN_INFLECTION`

(Added 2020-07-22) Similar to the `RATE_OF_INFLECTION` parameter froma bove, this value determines the rate of inflection from `LOCKDOWN_R_0` to `REOPEN_R`, as well as from `REOPEN_R` to `POST_REOPENING_EQUILIBRIUM_R` (see below). This value is usually between 0.15-0.35. The lower the value, the slower (and longer) the transitions. Hence, large countries usually have a lower value while US states typically have a value between 0.25-0.35. Read more about our post-reopening assumptions [here](https://covid19-projections.com/about/#post-reopening).

#### `POST_REOPENING_EQUILIBRIUM_R`

(Added 2020-07-19) This is the equilibrium R_t value after sufficient time has passed from the initial reopening. By default, this is set to 1, but can be adjusted. After accounting for immunity (which lowers R_t), the equilibrium R_t is likely between 0.85 and 1.

#### `FALL_R_MULTIPLIER`

(Added 2020-07-07) We are incorporating a potential for a [fall wave](https://covid19-projections.com/about/#fall-wave) as a result of school reopenings and the beginning of influenza season. This is the daily multiplier applied to the R value. As of July/August, because it is still too early to learn this value, this is sampled randomly from a triangular distribution with mode 1.001.